### PR TITLE
feat(publicChat):user should be able to chat [#187300111]

### PR DESCRIPTION
**Pull request by Mukunzijames**
What does this PR do?

user should be able to chat in public chat

Description of Task to be completed?

user should be able to chat live we used socket.io firstly you have to login before being able to chat

What are the relevant pivotal tracker/Trello stories?

id of the project

[#187300111]

Questions:

### CODE REVIEW

**Review by RWEMAREMY — APPROVED** [review]
good

**Review by musabehonore — APPROVED** [review]

File: src/app/chat/page.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  const messages = await axiosRequest('GET', `/chats`, '', true);
+  return messages.data.data;
+};
+
+const Page = () => {
+  const dispatch = useDispatch();
+  const socket = useRef<ReturnType<typeof io> | null>(null);
+  const [typingStatus, setTypingStatus] = useState('');
+  useEffect(() => {
+    socket.current = io(process.env.NEXT_PUBLIC_API_URL as string, {
```
Comment (fab-ryan) [inline src/app/chat/page.tsx]: setup of io must not  be included here

File: src/app/chat/page.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      auth: {
+        token: JSON.parse(localStorage.getItem('token') ?? '')
+      },
+      autoConnect: true
+    });
+    fetchMessages().then(data => {
+      dispatch(setChats(data));
+    });
```
Comment (fab-ryan) [inline src/app/chat/page.tsx:25]: please use the previous hooks created

File: src/app/chat/page.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    socket.current?.on('receiveMessage', data => {
+      dispatch(addChat(data));
+    });
+
+    socket.current?.on('sendUserId', getId => {
+      localStorage.setItem('userId', getId);
+    });
+    return () => {
+      socket.current?.disconnect();
+    };
```
Comment (fab-ryan) [inline src/app/chat/page.tsx:45]: use callbacck Hooks

Comment (fab-ryan) [reply src/app/chat/page.tsx:45]: insteady of useEffect Learn the different between useCallback and useEffect

File: src/app/layout.tsx [hunk 4]
```diff
@@ -26,10 +26,10 @@ export default function RootLayout({
       </head>
       <body>
         <Providers>
-          <div className="flex min-h-screen flex-col justify-between min-w-screen bg-main-100 text-main-400 h-auto relative pt-20 px-0">
+          <div className="flex min-h-screen flex-col justify-between min-w-screen bg-main-100 text-main-400 h-auto relative  px-0">
             <Suspense fallback={<PageLoading />}>
               {children}
-              <Footer />
+              {/* <Footer /> */}
```
Comment (fab-ryan) [inline src/app/layout.tsx]: Why did you disable this footer

File: src/components/Card.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import React, { PropsWithChildren } from 'react';
+import { cn } from 'tailwind-cn';
+
+const MessageCard = ({
+  children,
+  senderId
+}: PropsWithChildren<{ senderId: string }>) => {
+  const isSelf = senderId === localStorage.getItem('userId');
```
Comment (fab-ryan) [inline src/components/Card.tsx]: create helper that checks userId not Here also userId may be undefined this means I can get error

File: src/components/ChatForm.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+}) => {
+  const [message, setMessage] = useState('');
+
+  const dispatch = useAppDispatch();
+  const { loading, profile, status } = useAppSelector(
+    (state: RootState) => state.profile
+  );
+
+  useEffect(() => {
+    dispatch(fetchProfile());
```
Comment (fab-ryan) [inline src/components/ChatForm.tsx:22]: same as here Please use hooks for dispatch

File: src/components/ChatForm.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+          id="message"
+          className="w-full p-3 rounded-md border-2 border-none outline-none"
+          placeholder="type your message"
+          onChange={e => {
+            setMessage(e.target.value);
+            socket?.current?.emit('typing', true, profile?.data.name);
+            setTimeout(() => {
+              socket?.current?.emit('typing', false);
+            }, 1000);
+          }}
```
Comment (fab-ryan) [inline src/components/ChatForm.tsx]: create this as separate function

File: src/components/ChatList.tsx [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+export const ChatList = () => {
+  const chats = useAppSelector(({ chat }) => chat.chats);
+  return (
+    <div className="flex flex-col gap-4 pb-24">
+      {chats?.map((message, index) => (
+        <ChatMessage
+          key={index}
+          {...message}
+          isSelf={message.senderId === localStorage.getItem('userId')}
```
Comment (fab-ryan) [inline src/components/ChatList.tsx:14]: same as here that function are required

File: src/components/Footer/Footer.tsx [hunk 9]
```diff
@@ -1,127 +1,127 @@
...
+//                   <FaCircle />
+//                   {itm.name}
+//                 </li>
+//               ))}
+//           </ul>
+//         </div>
+//       </div>
+//       <nav className="w-full bg-main-200 text-xs sm:text-sm md:text-md flex justify-center text-main-400 items-center p-4 gap-2">
+//         © 2024 ATLP -29 Designed by TeamAlpha
+//       </nav>
```
Comment (fab-ryan) [inline src/components/Footer/Footer.tsx]: why the whole footer is commented

File: src/redux/slices/chatSlice.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+const chatsSlice = createSlice({
+  name: 'chats',
+  initialState,
+  reducers: {
+    setChats: (state, action) => {
+      state.chats = action.payload;
+    },
+    setLoading: (state, action) => {
+      state.loading = action.payload;
```
Comment (fab-ryan) [inline src/redux/slices/chatSlice.ts:32]: Loading states is not being used in pages

File: src/utils/requests.ts [hunk 11]
```diff
@@ -48,26 +48,26 @@ axiosInstance.interceptors.request.use(
...
+//       ) {
+//         window.location.href = '/login';
+//       }
+//     }
+//     if (error.response?.status === 403) {
+//       window.location.href = '/';
+//     }
+//     return Promise.reject(error);
+//   }
+// );
```
Comment (fab-ryan) [inline src/utils/requests.ts]: Guys why changing someone else codes??

**Review by fab-ryan — CHANGES_REQUESTED** [review]
Work on those comments you have repetitive implementation in whole pages.
also don't touch someone else's task

**Review by fab-ryan — APPROVED** [review]

**Pull request merged**